src/tools/wasm-ctor-eval.cpp

```diff
@@ -0,0 +1,445 @@
+/*
+ * Copyright 2017 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// Loads wasm plus a list of functions that are global ctors, i.e.,
+// are to be executed. It then executes as many of them as it can,
+// applying their changes to memory as needed, then writes it. In
+// other words, this executes code at compile time to speed up
+// startup later.
+//
+
+#include <memory>
+
+#include "pass.h"
+#include "support/command-line.h"
+#include "support/file.h"
+#include "support/colors.h"
+#include "wasm-io.h"
+#include "wasm-interpreter.h"
+#include "ast/memory.h"
+
+using namespace wasm;
+
+struct FailToEvalException {
+  std::string why;
+  FailToEvalException(std::string why) : why(why) {}
+};
+
+// We do not have access to imported globals
+class EvallingGlobalManager {
+  std::map<Name, Literal> globals;
+  std::set<Name> dangerouses;
+  bool sealed = false;
+
+public:
+  void addDangerous(Name name) {
+    dangerouses.insert(name);
+  }
+
+  void seal() {
+    sealed = true;
+  }
+
+  bool operator==(const EvallingGlobalManager& other) {
+    return globals == other.globals;
+  }
+  bool operator!=(const EvallingGlobalManager& other) {
+    return !(*this == other);
+  }
+
+  Literal& operator[](Name name) {
+    if (dangerouses.count(name) > 0) {
+      std::string extra;
+      if (name == "___dso_handle") {
+        extra = "\nrecommendation: build with -s NO_EXIT_RUNTIME=1 so that calls to atexit that use ___dso_handle are not emitted";
+      }
+      throw FailToEvalException(std::string("tried to access a dangerous (import-initialized) global: ") + name.str + extra);
+    }
+    if (sealed) {
+      if (globals.find(name) == globals.end()) {
+        throw FailToEvalException(std::string("tried to access missing global: ") + name.str);
+      }
+    }
+    return globals[name];
+  }
+
+  struct Iter {
```

Comment:
Nit: might as well name it `Iterator`, to be as explicit as possible? I assume it'll get `auto`ed everywhere anyway.
... Not that `Iter` is particularly hard to translate, but the full name is slightly more explicit? "Nit" for a reason.